File: Euler841.hs
{-# LANGUAGE BangPatterns #-}

-- Haskell solution.
-- Computes alternating shading area for {p/q} with inradius 1 using a
-- stable cosine-difference formula and parallel evaluation across Fibonacci pairs.

import Control.Concurrent (forkIO)
import Control.Concurrent.Chan (Chan, newChan, readChan, writeChan)
import Control.Monad (forM_, replicateM_, unless)
import Data.Array.IO (IOArray, newArray, getElems, writeArray)
import GHC.Conc (getNumProcessors, setNumCapabilities)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import Text.Printf (printf)

format10 :: Double -> String
format10 x = printf "%.10f" x

checkRounded10 :: String -> Double -> String -> IO ()
checkRounded10 name got expected = do
    let s = format10 got
    unless (s == expected) $ do
        hPutStrLn stderr $ "Validation failed: " ++ name ++ " = " ++ s
        exitFailure

kahanSum :: [Double] -> Double
kahanSum = go 0.0 0.0
  where
    go !s !c [] = s
    go !s !c (x:xs) =
        let !y = x - c
            !t = s + y
            !c' = (t - s) - y
        in go t c' xs

areaA :: Int -> Int -> Double
areaA p q
    | q < 1 = error "q must be >= 1"
    | p <= 2 * q = error "need p > 2q"
    | otherwise =
        let !dp = fromIntegral p :: Double
            !delta = pi / dp
            !sd = sin delta
            !cd = cos delta
            !tanDelta = sd / cd
            go :: Int -> Double -> Double -> Double -> Double -> Double -> Double
            go !k !sinK !cosK !sign !acc !c
                | k > q = acc
                | otherwise =
                    let !sinNext = sinK * cd + cosK * sd
                        !cosNext = cosK * cd - sinK * sd
                        !diff = sd / (cosK * cosNext)
                        !term = sign * diff
                        -- Kahan summation to reduce cancellation.
                        !y = term - c
                        !t = acc + y
                        !c' = (t - acc) - y
                    in go (k + 1) sinNext cosNext (-sign) t c'
            !sumDiff =
                if q < 2
                    then 0.0
                    else go 2 sd cd 1.0 0.0 0.0
            !inner = sumDiff - tanDelta
            !sign = if even q then 1.0 else -1.0
        in dp * sign * inner

fibs :: [Integer]
fibs = 1 : 1 : zipWith (+) fibs (drop 1 fibs)

fibPairs :: [(Int, Int)]
fibPairs =
    let fs = take 36 fibs  -- F1..F36
    in [ (fromIntegral (fs !! n), fromIntegral (fs !! (n - 2)))
       | n <- [3 .. 34] ]

validate :: IO ()
validate = do
    checkRounded10 "A(8,3)" (areaA 8 3) "9.9411254970"
    checkRounded10 "A(130021,50008)" (areaA 130021 50008) "10.9210371479"
    hPutStrLn stderr "Validation checkpoints passed."

worker :: Chan (Maybe (Int, Int, Int)) -> IOArray Int Double -> Chan () -> IO ()
worker taskChan outArr doneChan = do
    task <- readChan taskChan
    case task of
        Nothing -> return ()
        Just (idx, p, q) -> do
            let !val = areaA p q
            writeArray outArr idx val
            writeChan doneChan ()
            worker taskChan outArr doneChan

main :: IO ()
main = do
    procs <- getNumProcessors
    let workers = max 1 procs
    setNumCapabilities workers

    validate

    let pairs = fibPairs
        n = length pairs
    outArr <- newArray (0, n - 1) 0.0 :: IO (IOArray Int Double)
    taskChan <- newChan
    doneChan <- newChan

    replicateM_ workers (forkIO (worker taskChan outArr doneChan))

    forM_ (zip [0 ..] pairs) $ \(idx, (p, q)) ->
        writeChan taskChan (Just (idx, p, q))
    replicateM_ workers (writeChan taskChan Nothing)

    replicateM_ n (readChan doneChan)

    vals <- getElems outArr
    let total = kahanSum vals
    printf "%.10f\n" total

File: Euler841.cpp
#include <cmath>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

namespace {

struct Options {
    bool run_checkpoints = true;
};

bool parse_arguments(int argc, char** argv, Options& options) {
    for (int i = 1; i < argc; ++i) {
        const std::string arg(argv[i]);
        if (arg == "--skip-checkpoints") {
            options.run_checkpoints = false;
            continue;
        }
        std::cerr << "Unknown argument: " << arg << '\n';
        return false;
    }
    return true;
}

double area_A(const int p, const int q) {
    const double dp = static_cast<double>(p);
    const double delta = std::acos(-1.0) / dp;
    const double sd = std::sin(delta);
    const double cd = std::cos(delta);
    const double tan_delta = sd / cd;

    double sum_diff = 0.0;
    if (q >= 2) {
        int k = 2;
        double sin_k = sd;
        double cos_k = cd;
        double sign = 1.0;
        double corr = 0.0;  // Kahan compensation.
        while (k <= q) {
            const double sin_next = sin_k * cd + cos_k * sd;
            const double cos_next = cos_k * cd - sin_k * sd;
            const double diff = sd / (cos_k * cos_next);
            const double term = sign * diff;

            const double y = term - corr;
            const double t = sum_diff + y;
            corr = (t - sum_diff) - y;
            sum_diff = t;

            sin_k = sin_next;
            cos_k = cos_next;
            sign = -sign;
            ++k;
        }
    }

    const double inner = sum_diff - tan_delta;
    const double outside_sign = ((q % 2) == 0) ? 1.0 : -1.0;
    return dp * outside_sign * inner;
}

std::vector<std::uint64_t> fibonacci_sequence(const int count) {
    std::vector<std::uint64_t> fibs;
    fibs.reserve(static_cast<std::size_t>(count));
    fibs.push_back(1ULL);
    fibs.push_back(1ULL);
    for (int i = 2; i < count; ++i) {
        fibs.push_back(fibs[static_cast<std::size_t>(i - 1)] + fibs[static_cast<std::size_t>(i - 2)]);
    }
    return fibs;
}

double solve() {
    const std::vector<std::uint64_t> fibs = fibonacci_sequence(36);  // F1..F36
    double sum = 0.0;
    double corr = 0.0;
    for (int n = 3; n <= 34; ++n) {
        const int p = static_cast<int>(fibs[static_cast<std::size_t>(n)]);
        const int q = static_cast<int>(fibs[static_cast<std::size_t>(n - 2)]);
        const double term = area_A(p, q);
        const double y = term - corr;
        const double t = sum + y;
        corr = (t - sum) - y;
        sum = t;
    }
    return sum;
}

bool run_checkpoints() {
    const double a1 = area_A(8, 3);
    if (std::fabs(a1 - 9.9411254970) > 5e-11) {
        std::cerr << "Checkpoint failed: A(8,3)" << '\n';
        return false;
    }
    const double a2 = area_A(130021, 50008);
    if (std::fabs(a2 - 10.9210371479) > 5e-11) {
        std::cerr << "Checkpoint failed: A(130021,50008)" << '\n';
        return false;
    }
    return true;
}

}  // namespace

int main(int argc, char** argv) {
    Options options;
    if (!parse_arguments(argc, argv, options)) {
        return 1;
    }
    if (options.run_checkpoints && !run_checkpoints()) {
        return 2;
    }
    std::cout << std::fixed << std::setprecision(10) << solve() << '\n';
    return 0;
}
